AIRFLOW-2664: Support filtering dag runs by id prefix in API.

rcorre opened the pull request:
This allows users to choose run ids that will later allow them to filter
while retrieving runs.
Resolves part of AIRFLOW-2664.

Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/AIRFLOW-2664) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2664

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

`GET /dags/<dag_id>/dag_runs?run_id_prefix=<prefix>` will limit the
result to runs with an ID that starts with <prefix>.

For example, if you have a DAG that is used to provision namespaced
resources, you might name the runs like
`provision__$namespace__$timestamp`. You could then list runs for only a
given namespace using this query parameter.

This could be done client-side, but the result could be huge, so it is
useful to provide server-side filtering.

Filtering based on conf fields would be a more versatile solution, but I
didn't have an idea of how to cleanly do that through query parameters.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

- tests/www_rbac/api/experimental/test_dag_runs_endpoint.py:TestDagRunsEndpoint:test_get_dag_runs_success_with_prefix_parameter

I couldn't figure out how to run it locally. After following the steps in `CONTRIBUTING.md`, to set up a local dev environment:

```
virtualenv .env
. .venv/bin/activate
pip install -e .[devel]
./run_unit_tests.sh tests/www_rbac/api/experimental/test_dag_runs_endpoint.py:TestDagRunsEndpoint
```

the tests fail with:
```
ModuleNotFoundError: No module named 'unicodecsv'
```

I set up travis and it is still chugging away [here](https://travis-ci.org/rcorre/incubator-airflow).

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"


### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

    Added to the API function docstring, I assume this is sufficient?


### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Commit:
[AIRFLOW-2664] Add API dag_run list prefix param.

`GET /dags/<dag_id>/dag_runs?run_id_prefix=<prefix>` will limit the
result to runs with an ID that starts with <prefix>.

For example, if you have a DAG that is used to provision namespaced
resources, you might name the runs like
`provision__$namespace__$timestamp`. You could then list runs for only a
given namespace using this query parameter.

This could be done client-side, but the result could be huge, so it is
useful to provide server-side filtering.

Filtering based on conf fields would be a more versatile solution, but I
didn't have an idea of how to cleanly do that through query parameters.

Commit:
Use literal timestamp for trigger_dag in tests.

Tests for certain backends were failing when triggering more than one
dags from a test with "No row was found for one()". The failing code is
searching for a dag based on dag_id, run_id, and execution_date. I'm
hoping using a fixed execution_date, rather than defaulting to utcnow(),
will make the tests more predictable.

https://travis-ci.org/rcorre/incubator-airflow/jobs/396592289#L7127

ashb commented on airflow/models.py:
```diff
@@ -4860,6 +4860,8 @@ def find(dag_id=None, run_id=None, execution_date=None,
         :type run_id: string
         :param execution_date: the execution date
+        :param run_id_prefix: match runs with the given prefix in the run_id.
+        :type run_id_prefix: string
```
Ordering - you've split execution_date desc from it's type

ashb commented on airflow/models.py:
```diff
@@ -4849,7 +4849,7 @@ def refresh_from_db(self, session=None):
     @staticmethod
     @provide_session
-    def find(dag_id=None, run_id=None, execution_date=None,
+    def find(dag_id=None, run_id=None, execution_date=None, run_id_prefix=None,
```
Please add run_id_prefix (or whatever we decide to call the new param) to almost the end of the list, just before `session`

ashb commented on airflow/models.py:
```diff
@@ -4878,6 +4880,8 @@ def find(dag_id=None, run_id=None, execution_date=None,
             qry = qry.filter(DR.dag_id == dag_id)
         if run_id:
             qry = qry.filter(DR.run_id == run_id)
+        if run_id_prefix:
```
This one should probably be an `elif` as specifying run_id and run_id_prefix together doesn't make much sense

ashb commented on airflow/models.py:
```diff
@@ -5154,6 +5158,19 @@ def get_latest_runs(cls, session):
 
+    @classmethod
+    @provide_session
+    def get_runs(cls, dag_id, run_id_prefix, session=None):
```
Nothing calls this method does it? At the very least if it's useful to you then we need tests covering it, and more detailed docstring.

rcorre replied:
Nope, realized I could do it more easily by adding a param to `find` and forgot to remove this

Commit:
Fix code review comments for run_id filtering.

- Move run_id_prefix parameter
- run_id_prefix is ignored if run_id is passed
- Remove unused get_runs

Commit:
Replace run_id_prefix with run_id__like.

The latter is more flexible, as it allows users to provide any pattern.

Commit:
Fix fromtimestamp param in test_dag_runs_endpoint.

From timestamp needs a float, not a string.
